### ISSUE TITLE
General performance improvement on startup big projects

### ISSUE THREAD

**Issue by piotrzarzycki21** [prominic/Moonshine-IDE#756]
In past 6 months I had an opportunity to work on a Flex/Royale projects which contains more than 1000 files. I would like to put here results of discussion with @rat-moonshine and my experience with opening bigger projects Flex/Royale in Moonshine. 
Each opened project in Moonshine consume a lot of memory, depends on the size of project it fluctuates from 100MB to even more than 1GB, but memory is not a problem these days. However the CPU time is different thing. This is where Moonshine have some kind of problems and significantly not only slow or freeze, but also make some older machine problems to work smoothly. 

We have found two areas which may have influence over above problems:
1) Language server
I have project with around 20 modules/libs etc. I'm opening from that two projects which in summary has around 1000 files. Each language server instance takes around 500-700 MB of memory, but like I said the size doesn't matter here too much. Language server seems to be starting for both project quite fast, no problem here. However when LS starts next what is happening is that my CPU is 100% most of the time. It looks at some point that LS is doing something with project files, which takes a lot of time, sometimes even after half hour I do experience this problem.
Once I open project and LS tells me that it's started I do jump quick to file and wanted to do something, but it's usually super slow, cause I see that LS have it's increased activity - like despite of notification about complete start - it's doing something more - maybe scanning files ?

2) Iterating trough all files
We gave some mechanism in Moonshine which iterates trough all folders and gathers files to feed later Find Resources. This potentially can be one of the reason of such problems.

**Comment by piotrzarzycki21** [comment]
We are going to to improve soon **2** and see how it helps with this issue.

**Comment by rat-moonshine** [comment]
Thanks for creating the issue, @piotrzarzycki21 . I shall arrange some time to work on option 2 we discussed.

**Comment by rat-moonshine** [comment]
**_Tested on macOS Catalina (8GB RAM)_**

Opened projects:
- MoonshineDESKTOP
- MoonshineWEB
- Moonshine SharedCore
- MoonshineGUICore
- MoonshineWorker

With JDK/SDK properly setup | Without JDK path | Without file-reference recording on startup
---------------------------- | ------------------------ | -------------------------------
Language Server started</br>(2.5 Min) | Language Server do not starts<br/>(2.5 Min) | To be tested

**Comment by joshtynjala** [comment]
@piotrzarzycki21 Are you on macOS? If you are, you might be running into this issue: BowlerHatLLC/vscode-as3mxml#420 (and there's probably someone else with the same issue here: BowlerHatLLC/vscode-as3mxml#509)

Unfortunately, I haven't been able to reproduce on my Mac with any of my projects, and no one has been able to share a project with me yet.

**Comment by piotrzarzycki21** [comment]
> @piotrzarzycki21 Are you on macOS? If you are, you might be running into this issue: [BowlerHatLLC/vscode-as3mxml#420](https://github.com/BowlerHatLLC/vscode-as3mxml/issues/420) (and there's probably someone else with the same issue here: [BowlerHatLLC/vscode-as3mxml#509](https://github.com/BowlerHatLLC/vscode-as3mxml/issues/509))
> 
> Unfortunately, I haven't been able to reproduce on my Mac with any of my projects, and no one has been able to share a project with me yet.

Nope @joshtynjala - I'm on Windows.

**Comment by rat-moonshine** [comment]
Regarding the slow opening/importing of the big projects, as my investigation continues I have a growing feeling that the `FileWrapper.updateChildren()` is causing this slowness - which I also mentioned in my last comment to the issue. 

Now, this method also serves to populate the sidebar tree, so we can't just remove it. As I mentioned, I did include some way to not parse the whole file-system at a time but on-demand - I need to re-visit those codes, and therefore we probably also need to ensure that all other areas also work satisfactorily post the changes.

**Comment by JoelProminic** [comment]
We discussed using system commands to iterate through the project files instead of the AIR API.   If you just need to list the files and their directory, you can use commands like this:

```
# macOS
find %projectpath%
# Windows - needs more testing
cd %projectpath%; dir /a-d /b /s
```

If you need to collect more information about the files (like the timestamp), then these commands may be less useful (though the `find` command at least has a lot of options that could make this work).

**Comment by rat-moonshine** [comment]
I think a file-name is all that we require along with its path, in Find Resources. So the command return on macOS is all satisfactory. Its blazing fast while tested on Terminal.

**Comment by piotrzarzycki21** [comment]
> Regarding the slow opening/importing of the big projects, as my investigation continues I have a growing feeling that the `FileWrapper.updateChildren()` is causing this slowness - which I also mentioned in my last comment to the issue.
> 
> Now, this method also serves to populate the sidebar tree, so we can't just remove it. As I mentioned, I did include some way to not parse the whole file-system at a time but on-demand - I need to re-visit those codes, and therefore we probably also need to ensure that all other areas also work satisfactorily post the changes.

If you could make to that improvements any priority that would be good. My ability to test behavior on that big project may end up at the end of December.

**Comment by rat-moonshine** [comment]
Okay. Some updates here. 

When I met the first requirement (opening the project without a delay), I'm trying to address some rendering issues to the sidebar while opening Moonshine with projects. 

I shall update with the outcome shortly, hopefully.

**Comment by piotrzarzycki21** [comment]
Wen you have build for me ready. Let me know.

**Comment by rat-moonshine** [comment]
Okay. Here's a build for your tests @piotrzarzycki21 . Since you're mostly spending on your Windows workstation, I have uploaded a Windows build for you at here: https://app.box.com/s/r1dxgf62ubgr5xkbcspidigwb1dbdzyh.

Things you want to tests are:
1. Opening of the larger sized project(s) to the sidebar
2. Opening of Moonshine with larger sized projects opened to the sidebar

NOTE: You don't need to test any other functionality other than the above. It's possible some features might not work in an expected manner in this build, i.e. Find Resources.

For @JoelProminic , you can download the following builds from Bamboo for a test:
macOS: https://bamboo.prominic.net/browse/MS-MGOBB-299/artifact
Windows:

**Comment by piotrzarzycki21** [comment]
> NOTE: You don't need to test any other functionality other than the above. It's possible some features might not work in an expected manner in this build, i.e. Find Resources.

I need fully working IDE where I will be able to work. I don't want to install on clients machine something which I cannot use. I need to try it out for some time to  understand your improvements.

**Comment by rat-moonshine** [comment]
@piotrzarzycki21 a fully working IDE will require time as there are more areas which require to be adjusted, i.e. Find Resources. This build gave you to try/test only the slow-opening of projects locally, not in an intention to install on a client's machine. Unfortunately, I don't think I can help you with this part immediately.

On a separate note, besides the big project from your client, you can also test with the Moonshine projects locally, if these [test-values](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-734801067) match your workstation too, in general; You can test the updated opening time.

**Comment by joshtynjala** [comment]
I made some tweaks to how the ActionScript & MXML language server initializes. This may not affect project startup time very much, but what the user sees should more accurately reflect what's happening in the background. In particular, you will probably see the "Starting ActionScript & MXML code intelligence..." message for longer.

**Comment by rat-moonshine** [comment]
> In particular, you will probably see the "Starting ActionScript & MXML code intelligence..." message for longer.

Can we have a reference adds to the message against which of the opened project the language-server is starting currently?

**Comment by rat-moonshine** [comment]
For the performance improvements, I'm seeing three different areas where we can work on (we can have separate issues for each):

1. Opening time of the project(s), includes Moonshine startup time with project(s) opened
2. Tuning "Find Reference" file parsing as we discussed
3. Resource file copying process during a project's compilation; I noticed system freeze during resource copying while compiling Moonshine source in Moonshine-IDE. I imagine for a more smart system which can detect file-changes in specified resource-folders and copy those items only in consecutive steps (of course, full directory copying is obvious for first time compilation). We need to move these processes to shell scripts (or to any other speedy that may available) instead of using AIR File API.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine  Can you prepare build with all stuff which we have in master and Josh's updates ?

**Comment by rat-moonshine** [comment]
Yes, you can download an updated build with the master from here: https://app.box.com/s/wwykaxx0227qkj1vxkzozu5eydk30fv5. This includes the updates from @joshtynjala .

**Comment by piotrzarzycki21** [comment]
I have finally found time to launch that build. It's definitely if I don't count Language Server starts launching faster. I believe this build has still issue with JDK 11 -> https://github.com/prominic/Moonshine-IDE/issues/755#issuecomment-743434456 cause Language server didn't start for me on opened projects. 

I think you should prepare for me another one with all fixes and maybe than whole picture will be much better. Overall there is improvement.

Unfortunately Find Resources doesn't work - I mean I see there only couple of files instead all. Opened projects have around 1200 files.

**Comment by piotrzarzycki21** [comment]
More update. I have installed Adobe JDK 11 and start Moonshine again on clients PC. I have 4 Royale projects in Workspace. Three of them have at least 1300 files. Third is small couple of files. 

I would say that Moonshine had two phases of launching.
1. When it's opening the projects - that part seems to be much faster than it was before - this is improvement.
2. Starting a language server for all projects - it took around 5 minutes or more. During that operation Moonshine occasionally freezes. However after that operation it started work really great - looks like it faster than before.

@joshtynjala  is it possible to display on status bar what project language server is loading ? 

Unfortunately Find Resources problems remains - it doesn't display all files. @rat-moonshine If you could upgrade build with fresh changes I could test Moonshine for non Adobe JDK.

@rat-moonshine What would be your next steps on improvements now ? Those one which you did so far are very helpful.

**Comment by joshtynjala** [comment]
> @joshtynjala is it possible to display on status bar what project language server is loading ?

Yes, I'll add the project name to the message in January.

**Comment by rat-moonshine** [comment]
> More update. I have installed Adobe JDK 11 and start Moonshine again on clients PC. I have 4 Royale projects in Workspace. Three of them have at least 1300 files. Third is small couple of files.
> 
> I would say that Moonshine had two phases of launching.
> 
> 1. When it's opening the projects - that part seems to be much faster than it was before - this is improvement.
> 2. Starting a language server for all projects - it took around 5 minutes or more. During that operation Moonshine occasionally freezes. However after that operation it started work really great - looks like it faster than before.
> 
> @joshtynjala is it possible to display on status bar what project language server is loading ?
> 
> Unfortunately Find Resources problems remains - it doesn't display all files. @rat-moonshine If you could upgrade build with fresh changes I could test Moonshine for non Adobe JDK.
> 
> @rat-moonshine What would be your next steps on improvements now ? Those one which you did so far are very helpful.

Yeah, as I have mentioned "Find Resources" changes (and maybe some other areas) are yet to be implemented post these updates as #759 .

However, I may be going for another round of investigation for project opening improvement which may cover the "Find Resources" implementation, as well. So make no mistake, the current implementation of the performance improvements, are not final. You may be going to see more investigation for this subject.

**Comment by piotrzarzycki21** [comment]
Whenever you will have new build let me know. It was really good to have finally such a quick start. Language server was the second thing, but apart of that I felt a boost. If it would work everything I would probably use that build farther

**Comment by piotrzarzycki21** [comment]
@rat-moonshine Will you be able to look into that and progress in the next 2 weeks ? I'm not sure how long I will have an opportunity to try your changes on a real project.

**Comment by rat-moonshine** [comment]
I shall try. I just sidetracked with other immediate issues.

**Comment by rat-moonshine** [comment]
Some updates here. When the startup has shown some improvement I noticed a slight amount of delay populating the Find Resources list when there are many projects opened in the sidebar that contains a good amount of files. 

Currently, I'm experimenting with different ways/possibilities to reduce/eliminate that delay. Hopefully, I shall come up with an acceptable solution, soon.

**Comment by rat-moonshine** [comment]
Okay. I think it's ready for yours tests now. 

@piotrzarzycki21 please download from https://app.box.com/s/hfoyi9e2bpin3ntsf5sjg5bhyi291j0e (Windows). 
@JoelProminic , @Aszusz , you can download a build from Bamboo branch build plans:
Windows: https://bamboo.prominic.net/browse/MS-MGWBB-194/artifact
macOS: https://bamboo.prominic.net/browse/MS-MGOBB-339/artifact

**Comment by piotrzarzycki21** [comment]
@rat-moonshine Everything should work in that build ? I mean code completion etc ?

**Comment by Aszusz** [comment]
@rat-moonshine I'm switching to this build on Windows for my haxe development. Startup seems a lot faster, even for small project.

@piotrzarzycki21 Code completion still works for me.

**Comment by rat-moonshine** [comment]
Some updates. 

When earlier the list used to show like this:
![image](https://user-images.githubusercontent.com/24370870/107500452-2c86fd80-6bbc-11eb-9394-1436bf918f25.png)

it is now showing the list as in our stable releases:
![image](https://user-images.githubusercontent.com/24370870/107501026-e5e5d300-6bbc-11eb-991f-2a1b8e231b24.png)

And some other tuning. [Links](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-775861255) updated.

**Comment by JoelProminic** [comment]
I confirmed the shortened paths - I think this looks much better.  

I like the idea of updating the Total count while the Find Resources list is loading.  I think that this should be a little more obvious, though - I missed this label in my first couple tests.

There was an issue opening the Find Resources window on macOS, but that has been resolved already.

@rat-moonshine and I noticed an issue where we can't type text in the Find Resources prompt.  To get past this, we need to press tab until the search field has focus - we can't get focus using the mouse.  I saw this on both Windows and macOS.

If I close the Find Resources window before the list finishes loading, I get this error:

    : Error #1009
    : TypeError: Error #1009
    : 	at actionScripts.plugin.findResources::FindResourcesPlugin/updatesFilesInUI()
    : 	at actionScripts.plugin.findResources::FindResourcesPlugin/onParseCompleted()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at MethodInfo-27049()
    : 	at MethodInfo-15391()
    :
    : Click here to Report a Bug

This may be a separate issue, but I noticed problems with the close (x) button for the Find Resources prompt.  The clickable area does not seem to match the icon, or the transparent parts of the icon are not clickable.

**Comment by joshtynjala** [comment]
> I noticed an issue where we can't type text in the Find Resources prompt. To get past this, we need to press tab until the search field has focus - we can't get focus using the mouse. I saw this on both Windows and macOS.

I think that I remember that Piotr mentioned a similar issue recently. I'll look into it. I think that the FeathersUIWrapper class will need some more special handling for focus passing between Flex and Feathers UI.

**Comment by rat-moonshine** [comment]
During our meeting, @JoelProminic noticed that the following feature is not working in this branch build:
![image](https://user-images.githubusercontent.com/24370870/107604541-758b9000-6c56-11eb-9557-a4ba86043e10.png)

When probably I know why it's not working, to speed up things in consideration of this issue I require to choose some cost-effective ways, which also ended-up missing of particular information which is required for the above functionality. 

I'll see what I can do about it.

**Comment by rat-moonshine** [comment]
Some updates here:

- Placement and color change to make the counting label more obvious (in view), in Find Resource window
- Fixed exception throwing when Find Resource window dismissed before its complete load
- Fixed bug when the files-count shows unrealistic large number in Find Resource window (this shown when dismissing the window before its complete data load)

![image](https://user-images.githubusercontent.com/24370870/107604541-758b9000-6c56-11eb-9557-a4ba86043e10.png)
Initially I was thinking if not providing one certain information to the editor (`FileWrapper`) in new procedure, causing this nuisance. Upon investigating I saw the logic behind determining the location of the item in sidebar not depends on that information. 

I see I can able to have the above functionality work in some cases but it's unstable. This maybe not a related issue following the changes. We rather want to have this a separate issue once we merge this.

**Comment by piotrzarzycki21** [comment]
Did you update links so I could check ?

**Comment by JoelProminic** [comment]
The "show in sidebar" bug seems to be related to the updates in this issue.  I can reproduce the behavior like this:

1. Load a workspace with the Moonshine source and dependencies
2. Collapse all projects
3. Restart Moonshine
4. Edit > Find Resources - wait for the files to populate
5. Search for build.xml, and open VisualEditorConverterLib/build.xml
6. Click the target button.  The sidebar selection updates correctly
7. Edit > Find Resources
8. Search for build.xml, and open MoonshineSDKInstaller/build/build.xml
9. Click the target button.  BUG:  The MoonshineSDKInstaller project opens, but not the build directory, and the selected file does not change
10. Click the target button again.  BUG:  nothing changes
11. Edit > Find Resources
12. Search for build.xml, and open MoonshineSharedCore/src/elements/templates/projects/ActionScript Browser Project/build/build.xml.template
13. Click the target button.  BUG:  The MoonshineSharedCore project opens, but not the build directory, and the selected file does not change
14. Click the target button again.  BUG:  nothing changes

We suspect that this is because the sidebar entries are loading dynamically, which breaks the logic to open nested files.  So, one level of nesting worked in steps 4-6, but it failed on two levels of nesting in steps 7-10

I tried this with the latest nightly build, and the "show in sidebar" feature worked properly.

**Comment by JoelProminic** [comment]
The updates for Find Resources look good to me.  I think the placement of the total makes it easier to notice now:

![image](https://user-images.githubusercontent.com/24532731/107683450-2f8af800-6c67-11eb-82dd-d4cfb5a3370e.png)

**Comment by rat-moonshine** [comment]
Find Resources updated logic time-test, against the last-stable release logic:

Parse Logic | Projects Opened | Files Count | Seconds to Complete
--------- | ----------------- | ----------------- | -------------------------
Existing/Stable | InstallerSharedCore, MoonshineDESKTOP, MoonshineGUICore, MoonshineSharedCore, VisualEditorConverterLib | 18613 | 5
Branch build | InstallerSharedCore, MoonshineDESKTOP, MoonshineGUICore, MoonshineSharedCore, VisualEditorConverterLib | 18613 | 3

(Time may vary between the running system, especially at the time of starting the language-servers both the logic shown affected)

**Comment by rat-moonshine** [comment]
![image](https://user-images.githubusercontent.com/24370870/107604541-758b9000-6c56-11eb-9557-a4ba86043e10.png)
Locate File feature is now fixed. [Links](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-775861255) updated.

**Comment by JoelProminic** [comment]
Testing on macOS, Locate File works great for directories nested up to 3 levels deep.  For example:  `MoonshineSDKInstaller/build/build.xml`

However, If I try files that are nested deeper (for example, `MoonshineSharedCore/src/elements/templates/projects/ActionScript Browser Project/build/build.xml.template`), I get this:

![image](https://user-images.githubusercontent.com/24532731/107792301-974e4b00-6d1a-11eb-9ff9-d031b865f8cb.png)

I see that:
- Only two directories were opened
- The twistie is missing for the directories like `MoonshineSharedCore/src/actionScripts`, and I can't expand them further.  Refreshing the directory or project does not fix this.

**Comment by JoelProminic** [comment]
After @rat-moonshine did some more testing with this, we found that the above behavior only triggers if the "Show hidden files/folders" option is enabled in Moonshine Settings > General.  You'll need to restart Moonshine after changing the setting to see this behavior update.

In addition, I could reproduce the behavior for a new Royale project, but not for a Java project.  Both projects had hidden files, so I'm still not sure about the underlying cause here.

**Comment by rat-moonshine** [comment]
I see I'm a bit out of sight of this whole hidden files functionalities in Moonshine. During my tests with last stable v3.0.0, I see there was some contradictory concepts, or the things are confusing, or I simply do not understand how this suppose to work as an user.

Testing on Windows, I put a hidden file in a project's folder, 
![image](https://user-images.githubusercontent.com/24370870/107923866-18fedc80-6f98-11eb-9354-0b8b0396c1dc.png)

Set the global settings in Moonshine to show hidden files:
![image](https://user-images.githubusercontent.com/24370870/107923969-3f247c80-6f98-11eb-9108-b09faeae09a5.png)

Restarted Moonshine (just to re-ensure), but the file didn't shown up in the sidebar even when by the global setting, I was expecting it to appear.

At this point, the global settings to display hidden files seem to be a broken functionality to me.

Next I unselected "Show hidden files/folders" option in global settings. I noticed the context menu option to sidebar "Mark as hidden" toggles based on the global options value:

Show hidden files/folders | Mark as hidden
---------------------------- | ----------------
Selected | Displays in context-menu
Un-selected | Not available in context-menu

I don't know what is the underlying logic behind this option toggling in context-menu. While marking any file with "Mark as hidden" affects only after Moonshine restart:

Show hidden files/folders | Displays the file with "Mark as hidden" option
---------------------------- | --------------------
Un-selected | No
Selected | Yes

Now, as an user I'm confused what I'm expecting with the global "Show hidden files/folders" - 
- Does it suppose to toggle system-level hidden files?
- Does it suppose to toggle only Moonshine marked _hidden_ files?
- If both is suppose to be, then which one will get the priority

Looking into the [FileWrapper](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineSharedCore/src/actionScripts/valueObjects/FileWrapper.as#L213) codes, it seem like it reads by an individual project's settings only, possibly the files those have passed by the `.as3config` configuration. I'm not sure where the global setting intervene in this displaying process, though.

If someone has any knowledge on how this thing suppose to work, or an user is suppose to expect by using both the available options, please comment. I want have clear understanding before I go a fix for [this](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-778340630).

**Comment by rat-moonshine** [comment]
> 
> 
> After @rat-moonshine did some more testing with this, we found that the above behavior only triggers if the "Show hidden files/folders" option is enabled in Moonshine Settings > General. You'll need to restart Moonshine after changing the setting to see this behavior update.
> 
> In addition, I could reproduce the behavior for a new Royale project, but not for a Java project. Both projects had hidden files, so I'm still not sure about the underlying cause here.

This problem is fixed for now:

Show hidden files/folders | Results
---------------------------- | -------------
Enabled | Properly displays directories with expand icon
Disable | Same as above

**Comment by piotrzarzycki21** [comment]
If my memory serves well, logic for show/hiding files was designed for Visual Editor projects type. We wanted to avoid user's modification in VE XML files and force user to use UI editor instead XML itself to changing interfaces of designed applications. The question is with what kind of project do you open and test your changes ?

**Comment by rat-moonshine** [comment]
> The question is with what kind of project do you open and test your changes ?

"Show hidden files/folders" is broken for any type of projects.

Everyone, let me know if any other problem you found with this original issue update, before we consider for a merge next.

**Comment by piotrzarzycki21** [comment]
@rat-moonshine please merge with master and provide me new links for test.

**Comment by rat-moonshine** [comment]
I have updated the test-links [here](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-775861255). I didn't have a chance yet to test the updated build after the merge. Please, give a test.

**Comment by piotrzarzycki21** [comment]
I have run provided build and here observation. 

1. In my initial run I have opened Workspace where I had several small example projects. Flex, Royale and Haxe. I got immediately after run following errors:

![issues_env](https://user-images.githubusercontent.com/24554795/108352629-69b84480-71e7-11eb-8791-7be99cd9123e.png)

2. Language server for some projects run without the problems, but for one example just hang out. I have waited around 10 minutes.

![stuck](https://user-images.githubusercontent.com/24554795/108352732-8fdde480-71e7-11eb-86d5-48d5fccd801f.png)

3. I thought that maybe this DataGridExample is somehow problem, so I have closed other projects and restart Moonshine. This time I didn't get any errors **1** and project run correctly. I decided at this point switch to workspace where I had two Big Royale project more than 1k files. In the status bar I got information that language server is trying to run, but it looks like that:

![null](https://user-images.githubusercontent.com/24554795/108353328-4f329b00-71e8-11eb-82e0-3dd530003e17.png)

These projects had already opened couple of tabs. Once language server finished start I got several errors in console

![error_3218](https://user-images.githubusercontent.com/24554795/108353435-6f625a00-71e8-11eb-96ec-9c7d5782e9f6.png)

**Comment by rat-moonshine** [comment]
I assume you're not running a VD, so the temp-environment-setup error, is turning me worried now. 

For the other problems, as my earliest thought, try running the same workspace on last nightly build that you may have (having Josh's last updates) and see if the problem is reproducible there as well. So far I can say, the issue update has no direct involvement to the language-server process, it looks like the problems you're having all related to language-server startup process. 

I remember in one meeting during demonstration by @JoelProminic , one project's language-server or Java continue to running for a long time as shown to the status bar.

**Comment by piotrzarzycki21** [comment]
> I assume you're not running a VD

What VD is ?

**Comment by piotrzarzycki21** [comment]
> For the other problems, as my earliest thought, try running the same workspace on last nightly build that you may have (having Josh's last updates) and see if the problem is reproducible there as well.

I did that - I have run build from master and problems are not reproducible there.

**Comment by piotrzarzycki21** [comment]
> So far I can say, the issue update has no direct involvement to the language-server process, it looks like the problems you're having all related to language-server startup process.

Maybe your changes exposes some problems which did occur from time to time earlier.

**Comment by rat-moonshine** [comment]
If you can easily reproduce your test 1-2 again, and if it's not a private project, can you share DataGridExample from your workspace? Meanwhile, I also try to give a test at my end with today's build.

**Comment by piotrzarzycki21** [comment]
> If you can easily reproduce your test 1-2 again, and if it's not a private project, can you share DataGridExample from your workspace? Meanwhile, I also try to give a test at my end with today's build.

Sure here you have, but remember that this example didn't have problem with running on itself. It hangs when I run with several other small projects.
[DataGridExample.zip](https://github.com/prominic/Moonshine-IDE/files/6002673/DataGridExample.zip)

**Comment by rat-moonshine** [comment]
I able to reproduce the error mentioned in 1 and 2 during restarting my workspace with DataGridExample along with many other projects:

> 1. In my initial run I have opened Workspace where I had several small example projects. Flex, Royale and Haxe. I got immediately after run following errors:
> 2. Language server for some projects run without the problems, but for one example just hang out. I have waited around 10 minutes.

However, after a restart I didn't encountered any of the above errors anymore in my tests, so far. It's possible that 2 is caused because of 1. I need to investigate, but the unreliable nature of the alert/error may make the testing procedures more difficult.

**Comment by rat-moonshine** [comment]
I did some updates in trying to fix "Local environment setup failed[1]" error. A separate issue is created to track the progress as #785.

@piotrzarzycki21 please give this a shot from downloading from here https://app.box.com/s/36vx8vetmjev4fnkzlzslq03tevo1hu4, and see if you can reproduce this particular error in your many test attempts. 

@JoelProminic please download from the latest branch builds, I'm particularly eager to see how this work on the VD you test on, since the errors seen more obvious there.

**Comment by rat-moonshine** [comment]
I updated something on "Local environment setup failed" error/alert. Details [here](https://github.com/prominic/Moonshine-IDE/issues/785#issuecomment-783933840).

The test links updated at [here](https://github.com/prominic/Moonshine-IDE/issues/756#issuecomment-775861255).

During my tests I saw some problem with workspace project opening, in a short if I have 6 projects it opens 2/3 during Moonshine restart. This is not related to #785 update. I shall investigate this separately and create an issue accordingly.

**Comment by rat-moonshine** [comment]
This branch is now merged into the 'master'.

**Comment by JoelProminic** [comment]
This seems to be working in master.   

Language server is still very slow for large projects, but we should create a separate issue if we want to improve that logic.
